src/main/com/scalyr/api/query/QueryService.java

```
@@ -111,6 +155,78 @@ public LogQueryResult logQuery(String filter, String startTime, String endTime,
+    }
+  }
+
+  private static final long Y2000 =  946684800L * ScalyrUtil.NANOS_PER_SECOND;
+  private static final long Y2100 = 4102444800L * ScalyrUtil.NANOS_PER_SECOND;
+
+  /**
+   * Split `[startTime, endTime)` into chunks of at most `chunkSizeHours`, if `chunkSizeHours` is greater than zero and start/end are
+   * both parseable as longs using nanosecond precision;  otherwise returns the input unchanged.
+   */
+  public static Stream<Pair<String>> splitIntoChunks(String startTime, String endTime, int chunkSizeHours) {
+    try {
+      long min = Long.parseLong(startTime);
+      long max = Long.parseLong(endTime);
+
+      if (min >= Y2000 && min < Y2100 && max >= Y2000 && max < Y2100)
+        return splitIntoChunks(min, max, chunkSizeHours).map(longPair -> new Pair<>(Long.toString(longPair.a), Long.toString(longPair.b)));
+
+    } catch (Exception ignored) {
+    }
+
+    Logging.log(Severity.warning, "local/warning/unchunkedQuery", "Could not split [" + startTime + ", " + endTime + ") into chunks, executing as-is");
+
+    return Stream.of(new Pair(startTime, endTime));
+  }
+
+  /** Split `[start, end)` into `[start, start + chunk), [start + chunk, start + chunk * 2), ... [start + chunk * N, end)`. */
+  public static Stream<Pair<Long>> splitIntoChunks(final long start, final long end, final int chunkSizeHours) {
+    final long chunkNs = chunkSizeHours * 60 * 60 * ScalyrUtil.NANOS_PER_SECOND;
+    ArrayList<Pair<Long>> ret = new ArrayList<>();
+
+    for (int n = 0; n < 1000; n++) { // 1000 is just a short-circuit in case of weird input
+      long chunkStart  = start + n*chunkNs;
+      long chunkEnd = chunkStart + chunkNs;
+      ret.add(new Pair(chunkStart, Math.min(end, chunkEnd)));
+      if (chunkEnd >= end)
+        return ret.stream();
+    }
+
+    Logging.log(Severity.warning, "local/warning/unchunkedQuery", "Too many chunks for [" + start + ", " + end + "); executing as-is");
```

In the same spirit as throwing an exception if we're unable to parse start/end time for chunking, shall we throw an exception here as well?